### PR TITLE
[master] Implemented oneDNN Backward Adaptive Pooling kernel 

### PR DESCRIPTION
**Pull request by piotrwolinski-intel**
## Description ##
This PR introduces oneDNN backward adaptive pooling implementation.

### CODE REVIEW

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [centos-gpu]

File: src/operator/contrib/adaptive_avg_pooling.cc [hunk 1]
```diff
@@ -25,11 +25,13 @@
 // #include "elemwise_op_common.h"
 #include "../elemwise_op_common.h"
 #if MXNET_USE_ONEDNN == 1
+#include "../nn/dnnl/dnnl_base-inl.h"
 #include "../nn/dnnl/dnnl_pooling-inl.h"
 #endif  // MXNET_USE_ONEDNN
 
 #define START_IND(a, b, c) static_cast<int>(std::floor(static_cast<float>(a * c) / b))
 #define END_IND(a, b, c)   static_cast<int>(std::ceil(static_cast<float>((a + 1) * c) / b))
+#define DIV_ROUND_UP(a, b) (((a) + (b - 1)) / b)
```
Comment (bartekkuncer) [inline src/operator/contrib/adaptive_avg_pooling.cc:34]: Does a have to be in brackets?

File: src/operator/nn/dnnl/dnnl_pooling-inl.h [hunk 2]
```diff
@@ -28,10 +28,13 @@
 
 #include <dnnl.hpp>
 #include <utility>
+#include <vector>
 
 #include "../pooling-inl.h"
 #include "./dnnl_base-inl.h"
 
+#define DIV_ROUND_UP(a, b) (((a) + (b - 1)) / b)
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_pooling-inl.h:36]: Does a have to be in brackets?

File: src/operator/nn/dnnl/dnnl_pooling.cc [hunk 3]
```diff
@@ -233,17 +234,30 @@ void InitPoolingPrimitiveParams(const PoolingParam& param,
 dnnl::pooling_forward::primitive_desc GetPoolingFwdPdesc(const PoolingParam& param,
                                                          const bool is_train,
                                                          const dnnl::memory::desc& data_md,
-                                                         const dnnl::memory::desc& out_md) {
-  CHECK(param.kernel.ndim() == 1 || param.kernel.ndim() == 2 || param.kernel.ndim() == 3)
-      << "Not Implemented";
+                                                         const dnnl::memory::desc& out_md,
+                                                         const bool use_adaptive_pooling) {
+  CHECK((param.kernel.ndim() >= 1 && param.kernel.ndim() <= 3) || use_adaptive_pooling)
+      << "Not Implemented";  // to be changed
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_pooling.cc:240]: to be changed when?

File: src/operator/nn/dnnl/dnnl_pooling.cc [hunk 4]
```diff
@@ -357,21 +364,30 @@ DNNLPoolingBwd& GetPoolingBwd(const PoolingParam& param,
     auto dst_md   = dnnl::memory::desc(dst_dims, get_data_type(data_md), any);
 
     // fwd hint
-    auto fwd_pd = GetPoolingFwdPdesc(param, true, data_md, dst_md);
+    auto fwd_pd = GetPoolingFwdPdesc(param, true, data_md, dst_md, use_adaptive_pooling);
 
     // creat bwd desc
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_pooling.cc:369]: create?

File: src/operator/nn/dnnl/dnnl_pooling.cc [hunk 5]
```diff
@@ -384,22 +400,40 @@ DNNLPoolingBwd& GetPoolingBwd(const PoolingParam& param,
...
     return;
   }
 
+  const PoolingParam& param = nnvm::get<PoolingParam>(attrs.parsed);
+
+  const NDArray& out_grad  = inputs[0];
+  const NDArray* workspace = nullptr;
+  const NDArray* in_data   = nullptr;
+  if (DNNLRequireWorkspace(param)) {
+    // The first two elements are the gradient of the outputs in forward.
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_pooling.cc:418]:
```suggestion
    // The first two elements are the gradients of the outputs in forward.
```

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu, windows-gpu]

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [windows-gpu]

**Review by bartekkuncer — APPROVED** [review]
Looks good.

File: src/operator/nn/dnnl/dnnl_pooling-inl.h (hunk 2 above)
Comment (piotrwolinski-intel) [reply src/operator/nn/dnnl/dnnl_pooling-inl.h]: Not necessarily, changed for better readability.

File: src/operator/contrib/adaptive_avg_pooling.cc (hunk 1 above)
Comment (piotrwolinski-intel) [reply src/operator/contrib/adaptive_avg_pooling.cc]: Not necessarily, changed for better readability.

File: src/operator/nn/dnnl/dnnl_pooling.cc (hunk 4 above)
Comment (piotrwolinski-intel) [reply src/operator/nn/dnnl/dnnl_pooling.cc]: Yes, it was meant to be create.

File: src/operator/contrib/adaptive_avg_pooling.cc [hunk 6]
```diff
@@ -277,10 +344,35 @@ The pooling kernel and stride sizes are automatically chosen for desired output
...
+#else
+    .set_attr<nnvm::FInplaceOption>("FInplaceOption",
+                                    [](const NodeAttrs& attrs) {
+                                      return std::vector<std::pair<int, int>>();
+                                    })
+#endif
+#if MXNET_USE_ONEDNN == 1
+    .set_attr<bool>("TIsDNNL", true)
+    .set_attr<FComputeEx>("FComputeEx<cpu>", AdaptiveAvgPoolOpBackwardExCPU)
+#endif
```
Comment (bgawrych) [inline src/operator/contrib/adaptive_avg_pooling.cc:375]: Use only one MXNET_USE_ONEDNN condition

**Comment by piotrwolinski-intel** [comment]
@mxnet-bot run ci [centos-cpu, centos-gpu, windows-gpu]

**Review by bgawrych — APPROVED** [review]

**Review by mozga-intel — APPROVED** [review]

**Pull request merged**